spec/support/cmap.rb

```diff
@@ -0,0 +1,346 @@
+                        'reason' => event.reason,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionCheckoutStarted
+                      {
+                        'type' => 'ConnectionCheckOutStarted',
+                    }
+                    when Mongo::Monitoring::Event::ConnectionCheckoutFailed
+                      {
+                        'type' => 'ConnectionCheckOutFailed',
+                        'reason' => event.reason,
+                      }
+                   when Mongo::Monitoring::Event::ConnectionCheckedOut
+                      {
+                        'type' => 'ConnectionCheckedOut',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::ConnectionCheckedIn
+                      {
+                        'type' => 'ConnectionCheckedIn',
+                        'connectionId' => event.connection_id,
+                      }
+                    when Mongo::Monitoring::Event::PoolCleared
+                      {
+                        'type' => 'ConnectionPoolCleared',
+                        'address' => event.address,
+                      }
+                    else
+                      next
+                    end
+
+            events << event unless @ignore.include?(event['type'])
+            events
+          end
+        end
+      end
+
+      private
+
+      # Converts the options used by the Ruby driver to the spec test format.
+      def normalize_options(options)
+        (options || {}).reduce({}) do |opts, kv|
+         case kv.first
+          when :max_idle_time
+            opts['maxIdleTimeMS'] = (kv.last * 1000.0).to_i
+          when :max_pool_size
+            opts['maxPoolSize'] = kv.last
+          when :min_pool_size
+            opts['minPoolSize'] = kv.last
+          when :wait_queue_size
+            opts['waitQueueSize'] = kv.last
+          when :wait_queue_timeout
+            opts['waitQueueTimeoutMS'] = (kv.last * 1000.0).to_i
+          end
+
+          opts
+        end
+      end
+
+      # Converts the options given by the spec to the Ruby driver format.
+      def process_options(options)
+        (options || {}).reduce({}) do |opts, kv|
+          case kv.first
+          when 'maxIdleTimeMS'
+            opts[:max_idle_time] = kv.last / 1000.0
+          when 'maxPoolSize'
+            opts[:max_pool_size] = kv.last
+          when 'minPoolSize'
+            opts[:min_pool_size] = kv.last
+          when 'waitQueueSize'
+            opts[:wait_queue_size] = kv.last
+          when 'waitQueueTimeoutMS'
+            opts[:wait_queue_timeout] = kv.last / 1000.0
+          end
+
+          opts
+        end
+      end
+
+      # Places operations run by the non-main thread in the `thread_ops` field of the corresponding
+      # `start` operation.
+      def preprocess
+        until spec_ops.empty?
+          processed_ops << spec_ops.shift
```

Comment:
I expect `processed_ops` to contain the result of execution, not be a copy of `spec_ops`. What does this line do?